ledger/src/blockstore/blockstore_purge.rs

```diff
@@ -144,11 +145,67 @@ impl Blockstore {
         self.run_purge_with_stats(from_slot, to_slot, purge_type, &mut PurgeStats::default())
     }
 
+    /// Purges all columns relating to `slot`.
+    ///
+    /// Additionally we cleanup the parent of `slot`, by clearing `slot` from
+    /// the parent's `next_slots`. We reinsert an orphaned `slot_meta` for `slot`
+    /// that preserves `slot`'s `next_slots`. This ensures that `slot`'s fork is
+    /// replayable upon repair of `slot`.
+    pub(crate) fn purge_slot_cleanup_chaining(&self, slot: Slot) -> Result<bool> {
+        let Some(mut slot_meta) = self.meta(slot)? else {
+            return Err(BlockstoreError::SlotUnavailable);
+        };
+        let mut write_batch = self
+            .db
+            .batch()
+            .expect("Database Error: Failed to get write batch");
+
+        let columns_purged = self.purge_range(&mut write_batch, slot, slot, PurgeType::Exact)?;
+
+        if let Some(parent_slot) = slot_meta.parent_slot {
+            let parent_slot_meta = self.meta(parent_slot)?;
+            if let Some(mut parent_slot_meta) = parent_slot_meta {
+                // .retain() is a linear scan; however, next_slots should
+                // only contain several elements so this isn't so bad
+                parent_slot_meta
+                    .next_slots
+                    .retain(|&next_slot| next_slot != slot);
+                write_batch.put::<cf::SlotMeta>(parent_slot, &parent_slot_meta)?;
+            } else {
+                // As lowest cleanup slot is not dependable, we cannot differentiate if the
+                // the parent was actually cleaned up. Once it is dependable, we can hard
+                // unwrap after a comparison with the parent.
+                error!(
+                    "Parent slot meta {} for child {} is missing  or cleaned up.
+                       Falling back to orphan repair to remedy the situation",
+                    parent_slot, slot
+                );
+            }
+        }
+
+        // Reinsert parts of `slot_meta` that are important to retain, like the `next_slots` field.
+        slot_meta.clear_unconfirmed_slot();
+        write_batch.put::<cf::SlotMeta>(slot, &slot_meta)?;
+
+        if let Err(e) = self.db.write(write_batch) {
+            error!(
+                "Error: {:?} while submitting write batch for slot {:?} retrying...",
+                e, slot
+            );
+            return Err(e);
+        }
```

Comment:
nit: I've hit this case before where I wanted to print a log message locally but still raise the error. After looking at docs real quick, it looks like the following might simplify this (new feature in Rust 1.76):
```rust
self.db.write(write_batch)
    .inspect_err(|e| error!("Error {:?} ....))?;
```
Also, I don't think that `retrying` bit in the message is valid here (or in the other spot) so can you please remove it

Reply:
nice, `inspect_err` is slick.